File: Data Structures & Algorithms/find-the-town-judge/submission-1.cpp
class Solution {
public:
    int findJudge(int n, vector<vector<int>>& trust) {
        std::unordered_map<int,std::vector<int>> outgoing;
        std::unordered_map<int,std::vector<int>> incoming;
        for (auto& vec:trust){
            outgoing[vec[0]].push_back(vec[1]);
            incoming[vec[1]].push_back(vec[0]);
        }
        std::unordered_set<int> c1;
        std::unordered_set<int> c2;
        std::unordered_set<int> res;
        for (int i = 1; i < n+1; ++i){
            if (!outgoing.contains(i)){
                c1.insert(i);
            }
        }
        for (int i = 1; i < n+1; ++i){
            if (incoming[i].size() == n-1){
                c2.insert(i);
            }
        }
        std::set_intersection(c1.begin(),c1.end(),c2.begin(),c2.end(),std::inserter(res,res.begin()));
        if (res.size() == 1){
            return *res.begin();
        }
        return -1;
    }
};

File: Data Structures & Algorithms/count-subsequences/submission-1.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        dp = [[0] * (len(s)+1) for i in range(len(t)+1)]
        for j in range(len(dp[0])):
            dp[-1][j] = 1
        for i in range(len(dp)-2,-1,-1):
            for j in range(len(dp[0])-2,-1,-1):
                if s[j] == t[i]:
                    dp[i][j] += dp[i+1][j+1]
                dp[i][j] += dp[i][j+1]
        return dp[0][0]
        # def dfs(s,t):
        #     if len(t) == 0:
        #         return 1
            
        #     res = 0
        #     if s:
        #         if s[0] == t[0]:
        #             res += dfs(s[1:],t[1:])
        #         res += dfs(s[1:],t)
        #     return res
        
        # return dfs(s,t)